R-package/configure.ac

```
@@ -21,7 +21,11 @@ if test -z "${R_HOME}"; then
     exit 1
 fi
 CC=`"${R_HOME}/bin/R" CMD config CC`
-CXX=`"${R_HOME}/bin/R" CMD config CXX11`
+CXX11=`"${R_HOME}/bin/R" CMD config CXX11`
+CXX11STD=`"${R_HOME}/bin/R" CMD config CXX11STD`
+CXX="${CXX11} ${CXX11STD}"
+CPPFLAGS=`"${R_HOME}/bin/R" CMD config CPPFLAGS`
+CXXFLAGS=`"${R_HOME}/bin/R" CMD config CXX11FLAGS`
```

I believe we should add the following line here (and in `configure.win` as well btw) according to the example in https://cran.r-project.org/doc/manuals/R-exts.html#Configure-and-cleanup.
```
Packages written with C++ need to pick up the details for the C++ compiler and switch the current language to C++ by something like

CXX=`"${R_HOME}/bin/R" CMD config CXX`
if test -z "$CXX"; then
  AC_MSG_ERROR([No C++ compiler is available])
fi
CXXFLAGS=`"${R_HOME}/bin/R" CMD config CXXFLAGS`
CPPFLAGS=`"${R_HOME}/bin/R" CMD config CPPFLAGS`
AC_LANG(C++)
```
The same line appears in https://cran.r-project.org/doc/manuals/R-exts.html#Using-C_002b_002b-code.


```suggestion
CXXFLAGS=`"${R_HOME}/bin/R" CMD config CXX11FLAGS`
AC_LANG(C++)
```

---

`AC_LANG(C++)` is already included in all the relevant places in `configure.ac`.

https://github.com/microsoft/LightGBM/blob/86ead2050c006b984e88396e0e380304bac114d1/R-package/configure.ac#L41

https://github.com/microsoft/LightGBM/blob/86ead2050c006b984e88396e0e380304bac114d1/R-package/configure.ac#L68

I guess it could be pulled up to this line to slightly reduce duplication, but that won't change the behavior of `configure` at all.

I don't think it's appropriate for `configure.win`. `AC_LANG` is an `autoconf` directive ([see here](https://www.gnu.org/software/autoconf/manual/autoconf-2.63/html_node/Language-Choice.html)), and `configure.win` is not used as an input to `autoconf` the way `configure.ac` is. `configure.win` is a regular shell script.

---

I've done that consolidation in `configure.ac` in https://github.com/microsoft/LightGBM/pull/4506/commits/70fcec0224d409520ed63fc4b7d77097a5b903ab.

---

> I don't think it's appropriate for `configure.win`

Ah, indeed, you're right!